av.c

```diff
@@ -393,6 +393,39 @@ Perl_av_store(pTHX_ AV *av, SSize_t key, SV *val)
     return &ary[key];
 }
 
+/*
+=for apidoc av_new_alloc
```

Comment:
As well as the implementation of `zeroflag`, this has some other differences to the newAV+av_extend equivalent that I think should either be corrected or documented. I notice at least:
- lower bound check (size < 0)
- upper bound check (memwrap)
- defaulting to minimum (max(size, 4))

My first guess is that the lack of default minimum should be documented, but the bounds checks should be added.

Reply:
Yes, I had intended that people using this are certain that they will be calling with a known, non-zero size. I will improve the documentation and add an assert for debugging builds.

I thought that Newx - i.e. safesysmalloc - checks for memwrap?

Reply:
> Yes, I had intended that people using this are certain that they will be calling with a known, non-zero size. I will improve the documentation and add an assert for debugging builds.

Ok.

> I thought that Newx - i.e. safesysmalloc - checks for memwrap?

Ah yes, you're quite right - so the `MEM_WRAP_CHECK_s` in `av_extend_guts` is actually a duplicate, presumably only for the purpose of giving a more specific error message. Maybe that could be combined into a `Newx_s` macro to avoid the duplication.

Reply:
Huh, _safesysmalloc_ does not check for memwrap. I think I intended that people using this function are certain that they are using a non-zero, positive integer that is small enough to never wrap! But I'd initially envisaged the usage being for small, hardcoded sizes, which is obviously not the case for the pp_hot.c function changed in this commit.

I'll add a check for memwrap.

Reply:
D'oh, I must have looked more closely the first time around! Let me look again.

Reply:
Yes, I misremembered where the memwrap check was. (i.e. The `Newx` macros, not the `safesysmalloc` function.)

So I propose to not add an extra check into this PR and we open a separate issue for the `Newx_s` macro you suggest. Sound okay?

Reply:
> So I propose to not add an extra check into this PR and we open a separate issue for the `Newx_s` macro you suggest. Sound okay?

Yes.